### PR TITLE
Introduce helpers that check if 2 eventuals can be composed

### PR DESCRIPTION
**Pull request by ArthurBandaryk**

### CODE REVIEW

File: eventuals/event-loop.h [hunk 1]
```diff
@@ -422,6 +422,11 @@ class EventLoop final : public Scheduler {
             Errors,
             std::tuple<std::runtime_error>>;
 
+        template <typename Downstream>
+        static constexpr bool CanCompose = Downstream::ExpectsValue;
+
+        using Expects = SingleValue;
```
Comment (FolMing) [inline eventuals/event-loop.h:428]: Timer's Start function takes void as argument. `using Expects = Nothing`?

Comment (benh) [reply eventuals/event-loop.h:428]: `void` is a "single value"!

File: eventuals/compose.h [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <iostream>
```
Comment (benh) [inline eventuals/compose.h:3]:
```suggestion
```

File: eventuals/event-loop.h [hunk 3]
```diff
@@ -1032,6 +1042,11 @@ class EventLoop final : public Scheduler {
           Errors,
           std::tuple<std::runtime_error>>;
 
+      template <typename Downstream>
+      static constexpr bool CanCompose = Downstream::ExpectsStream;
+
+      using Expects = StreamOfValues;
```
Comment (benh) [inline eventuals/event-loop.h:1048]:
```suggestion
      using Expects = SingleValue;
```

File: eventuals/generator.h [hunk 4]
```diff
@@ -472,6 +472,11 @@ struct _Generator final {
           std::move(dispatch_));
     }
 
+    template <typename Downstream>
+    static constexpr bool CanCompose = true;
+
+    using Expects = StreamOrValue;
```
Comment (benh) [inline eventuals/generator.h:478]: Unless I'm missing something this is:
```suggestion
    using Expects = SingleValue;
```

File: eventuals/map.h [hunk 5]
```diff
@@ -134,6 +134,11 @@ struct _Map final {
       }
     }
 
+    using Expects = StreamOfValues;
+
+    template <typename Downstream>
+    static constexpr bool CanCompose = Downstream::ExpectsStream;
```
Comment (benh) [inline eventuals/map.h:140]:
```suggestion
    template <typename Downstream>
    static constexpr bool CanCompose = Downstream::ExpectsStream;

    using Expects = StreamOfValues;
```

File: eventuals/range.h [hunk 6]
```diff
@@ -91,6 +91,11 @@ struct _Range final {
       return Continuation<K, Arg>(std::move(k), from_, to_, step_);
     }
 
+    template <typename Downstream>
+    static constexpr bool CanCompose = Downstream::ExpectsStream;
+
+    using Expects = StreamOfValues;
```
Comment (benh) [inline eventuals/range.h:97]:
```suggestion
    using Expects = SingleValue;
```

File: eventuals/repeat.h [hunk 7]
```diff
@@ -73,6 +73,11 @@ struct _Repeat final {
     auto k(K k) {
       return Continuation<K>(std::move(k));
     }
+
+    template <typename Downstream>
+    static constexpr bool CanCompose = Downstream::ExpectsStream;
+
+    using Expects = StreamOfValues;
```
Comment (benh) [inline eventuals/repeat.h:80]:
```suggestion
    using Expects = SingleValue;
```

File: eventuals/terminal.h [hunk 8]
```diff
@@ -153,6 +153,10 @@ struct _Terminal final {
           std::move(stop));
     }
 
+    template <typename Downstream>
+    static constexpr bool CanCompose = Downstream::ExpectsValue;
```
Comment (benh) [inline eventuals/terminal.h:157]:
```suggestion
    static constexpr bool CanCompose = false;
```

File: eventuals/transformer.h [hunk 9]
```diff
@@ -240,6 +240,11 @@ struct _Transformer final {
     template <typename Arg, typename Errors>
     using ErrorsFrom = tuple_types_union_t<Errors, Errors_>;
 
+    template <typename Downstream>
+    static constexpr bool CanCompose = true;
```
Comment (benh) [inline eventuals/transformer.h:244]:
```suggestion
    static constexpr bool CanCompose = Downstream::ExpectsStream;
```

**Review by benh — APPROVED** [review]

**Pull request merged**